### PR TITLE
fix(printer): generalize evidence redaction beyond Secret-kind-only

### PR DESCRIPTION
**Pull request by adity1raut**
## Summary
- `isSensitivePath()` only masked a path's value when the resource kind was exactly `Secret` and the path was under `data`/`stringData`. Any other resource storing a credential directly - a ConfigMap entry named `apiKey`, a CRD's `spec.auth.token`, a field named `clientSecret` - had its value printed in the clear via the `(current: ...)` annotation, since the check never looked past `kind == "Secret"`.
- Add a kind-independent check: if a path's final field name looks like a credential (`password`, `token`, `apiKey`, `clientSecret`, `db_password`, ...; matched case-insensitively with separators normalized), mask it regardless of resource kind.
- The pattern list mirrors `anonymizer.isSensitiveEnvName`'s list in `core/pkg/anonymizer`, kept local rather than imported since `anonymizer` imports `resultshandling`, which imports this package - importing `anonymizer` here would create an import cycle.
- This intentionally does not correlate a generic field (e.g. a container env var's `value`) with a sibling field that names it (e.g. that same env var's `name`) - path-based matching alone can't see that relationship, so a hardcoded secret sitting in an env var's value is still out of reach of this check.

## Test plan
- [x] `go test ./core/pkg/resultshandling/printer/v2/...`
- [x] `go vet ./core/pkg/resultshandling/printer/v2/...`
- [x] `gofmt -l` clean
- [x] `go build ./...`
- [x] Added cases to `TestIsSensitivePath` covering cross-kind matches (ConfigMap, custom resource, snake_case/uppercase normalization) and a negative case (`spec.keyword`) confirming no false-positive substring match against `apikey`/`accesskey`

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved credential redaction based on Kubernetes resource type and field location.
  * Prevented similarly named fields in custom resources from being incorrectly treated as safe.
  * Reduced false-positive masking for documented fields such as `secretName`, `serviceAccountToken`, and `automountServiceAccountToken`.
  * Continued protecting genuinely sensitive values, including secrets, tokens, Secret data, and container environment values.

### CODE REVIEW

File: core/pkg/resultshandling/printer/v2/pathvalue.go [hunk 1]
```diff
@@ -120,20 +120,74 @@ func extractValueAtPath(obj map[string]any, path string) (string, bool) {
...
+	segments := splitPath(path)
+	if len(segments) == 0 {
+		return false
+	}
+	name := strings.ToLower(segments[len(segments)-1].key)
+	for _, sep := range []string{"_", "-", ".", " "} {
+		name = strings.ReplaceAll(name, sep, "")
+	}
+	for _, pattern := range secretFieldPatterns {
+		if strings.Contains(name, pattern) {
```
Comment (matthyx) [inline core/pkg/resultshandling/printer/v2/pathvalue.go:334]: Substring match (`strings.Contains`) on the fully-concatenated, separator-stripped name means `automountServiceAccountToken` matches `"token"` and gets treated as secret-shaped even though it's an unrelated boolean field. This is what breaks `TestReviewPathsWithCurrentValues/value_extracted` in CI. Consider matching against whole words split at camelCase/snake_case/kebab-case boundaries instead of a raw substring check on the concatenated name.

**Review by matthyx — CHANGES_REQUESTED** [review]
Nice fix for the Secret-kind-only blind spot, and good call keeping the pattern list local to avoid the `anonymizer` import cycle — I confirmed it mirrors `anonymizer.isSensitiveEnvName` exactly.

There's a real regression though: CI's `pr-scanner / Create cross-platform build` job is failing (https://github.com/kubescape/kubescape/actions/runs/31893831928/job/95033967668) — `go test ./core/pkg/resultshandling/printer/v2/...` fails on the pre-existing `TestReviewPathsWithCurrentValues/value_extracted` test:

```
-spec.automountServiceAccountToken (current: true)
+spec.automountServiceAccountToken
```

Root cause: `hasSecretShapedFieldName` uses plain `strings.Contains(name, pattern)` after stripping separators, so it's a raw substring match rather than a word-boundary match. `automountServiceAccountToken` normalizes to `automountserviceaccounttoken`, which contains `"token"` as a substring even though the field is an unrelated boolean flag, not a credential. So this boolean gets its `(current: true)` annotation stripped, and (worse from the security-review angle this PR cares about) any resource with a `false` value there now also gets falsely redacted — not what you want.

This is likely to bite other real K8s field names too, since substring matching on unnormalized-camelCase-turned-lowercase is fairly loose, e.g. `secretName` (a *reference* to a Secret by name, not the value itself), `tokenExpirationSeconds` / `serviceAccountToken` (projected volume fields), etc.

Could you switch to a word-aware match — e.g. split the field name into its constituent words (camelCase/snake_case/kebab-case boundaries) and check for exact word membership in `secretFieldPatterns`, rather than `strings.Contains` on the fully-concatenated name? And please add `spec.automountServiceAccountToken` (and similar "name contains the substring but isn't a secret" cases) as explicit regression tests in `TestIsSensitivePath` so this doesn't reappear.

Once the CI failure is fixed and the false-positive is addressed, this looks good to merge — the underlying idea (kind-independent path-based redaction) is a solid improvement.

**Comment by adity1raut** [comment]
Fixed the CI failure — thanks for catching it, and for the specific repro.

I ended up not implementing word-boundary splitting as suggested, and wanted to explain why rather than silently diverging: I traced through it and "Token"/"Secret" are genuine, complete words in `automountServiceAccountToken`, `serviceAccountToken`, and `secretName` — not substring artifacts spanning two unrelated words (e.g. it's not like "cookie" containing "poke"). A word-boundary-aware exact-match would still flag all three, since the credential-shaped word really is a standalone word there. This is a semantic distinction (a field *naming or describing* a credential vs. a field that *holds* one) rather than a tokenization problem, so I added a small, explicitly-documented `safeFieldNames` exclusion list for these specific known Kubernetes API fields instead, checked before the existing substring match.

Added regression tests for `automountServiceAccountToken` (including a snake_case variant), `serviceAccountToken`, and `secretName`, plus confirmation that fields genuinely named `token`/`secret` are still caught. `TestReviewPathsWithCurrentValues/value_extracted` (the one failing in CI) now passes.

Happy to extend the exclusion list if you know of other real K8s fields in the same shape — I verified each of these three against the actual `k8s.io/api` type definitions rather than guessing (that's also why I dropped `tokenExpirationSeconds` and `configMapName` from my first draft — neither is an actual field name in the core API; the real fields are `expirationSeconds` and `name` respectively).

File: core/pkg/resultshandling/printer/v2/pathvalue.go [hunk 2]
```diff
@@ -120,20 +120,103 @@ func extractValueAtPath(obj map[string]any, path string) (string, bool) {
...
+	if len(segments) == 0 {
+		return false
+	}
+	name := strings.ToLower(segments[len(segments)-1].key)
+	for _, sep := range []string{"_", "-", ".", " "} {
+		name = strings.ReplaceAll(name, sep, "")
+	}
+	if _, safe := safeFieldNames[name]; safe {
+		return false
+	}
```
Comment (matthyx) [inline core/pkg/resultshandling/printer/v2/pathvalue.go:332]: The word-boundary fix is a good, well-reasoned response to the earlier CI blocker — `automountServiceAccountToken`/`serviceAccountToken`/`secretName` really are complete words, so a tokenizer alone wouldn't have distinguished them either, and the exclusion is clearly documented.

One remaining gap, same shape as the bug this PR set out to fix: `safeFieldNames` is keyed purely on the normalized field name, independent of `kind` (line 191-193). That means `isSensitivePath("SomeCustomResource", "spec.serviceAccountToken")` also returns `false` even if that CRD stores a real token value under that exact field name rather than a `ServiceAccountTokenProjection` block — silently un-redacting it. The reasoning in the comment above (`Token`/`Secret` are genuine words, not substring artifacts) is right for the specific built-in K8s types these exceptions were verified against, but it doesn't hold once a third-party CRD reuses the same field name for something else. Given the whole point of this change is kind-independent detection, could you gate these three exceptions on the (kind, path) combos where they're actually verified-safe (e.g. `Pod`/`PodSpec` for the first two, `SecretVolumeSource` shapes for `secretName`), falling back to the pattern match otherwise? A regression test with a non-core kind reusing one of these names (e.g. `kind: "MyCustomResource", path: "spec.serviceAccountToken"`) would pin the intended behavior.

File: core/pkg/resultshandling/printer/v2/pathvalue_test.go [hunk 3]
```diff
@@ -300,6 +300,23 @@ func TestIsSensitivePath(t *testing.T) {
...
+		{name: "keyword is not a false positive for key/apikey", kind: "Deployment", path: "spec.keyword", want: false},
+		// Regression: automountServiceAccountToken is a boolean toggle, not a
+		// token value - "Token" is a genuine word in it, so it previously
+		// matched the "token" pattern as a plain substring.
+		{name: "automountServiceAccountToken boolean is not a token value", kind: "Pod", path: "spec.automountServiceAccountToken", want: false},
+		{name: "automountServiceAccountToken is unaffected by case/separators", kind: "Pod", path: "spec.automount_service_account_token", want: false},
+		{name: "serviceAccountToken projected volume source is not a token value", kind: "Pod", path: "spec.volumes[0].projected.sources[0].serviceAccountToken", want: false},
+		{name: "secretName is a reference to a Secret object, not its value", kind: "Pod", path: "spec.volumes[0].secret.secretName", want: false},
+		{name: "a field that is actually named token is still caught", kind: "MyCustomResource", path: "spec.auth.token", want: true},
+		{name: "a field that is actually named secret is still caught", kind: "OAuthClient", path: "spec.clientSecret", want: true},
```
Comment (matthyx) [inline core/pkg/resultshandling/printer/v2/pathvalue_test.go]: Minor: this case's path is `spec.clientSecret`, which duplicates the `clientSecret is caught` case a few lines up — it doesn't actually exercise a field literally named `secret`. Suggest `spec.secret` here so the two cases cover distinct inputs.

**Review by matthyx — CHANGES_REQUESTED** [review]
Checked back in on this — the CI-breaking substring bug from my last review is genuinely fixed, and the `safeFieldNames` approach plus its regression tests (including the snake_case variant) is a solid, well-explained response to why word-boundary splitting alone wouldn't have worked. `automountServiceAccountToken` no longer loses its `(current: ...)` value.

Two things before this is ready to merge:

1. **Real gap, same shape as the original bug**: the new `safeFieldNames` exclusion is applied by field name alone, independent of `kind`. A non-core resource (CRD, etc.) that happens to reuse `serviceAccountToken` or `secretName` for an actual credential value would now silently skip redaction — left an inline comment with a suggested fix (gate the exceptions on kind/path, not name alone) and CodeRabbit flagged the same thing as a Major/CWE-200 finding independently, so this isn't just my read.
2. **Merge conflict**: GitHub currently reports this branch as unmergeable (`mergeable_state: dirty`) against `master` — needs a rebase/resolve regardless of the above.

Also left a small test nit (duplicate `clientSecret` path under a case meant to test a literal `secret` field name).

Happy to approve once the kind-scoping is addressed and the branch is rebased — the underlying feature is a real improvement and close to done.

**Comment by matthyx** [comment]
Checked again — no new commits addressing the open items since my last review (only a rebase against `master` landed, which resolves the merge-conflict blocker and CI is green). Still open: (1) `safeFieldNames` in `pathvalue.go` excludes by field name alone with no kind/path scoping, so a CRD reusing `serviceAccountToken`/`secretName` for a real credential would silently skip redaction — same gap CodeRabbit flagged as Major/CWE-200; (2) the test case at `pathvalue_test.go:391` still uses `spec.clientSecret`, duplicating the case above it instead of testing a literal `secret` field. Will approve once those are addressed.

**Comment by matthyx** [comment]
Checked again — both prior blockers are fixed in `7775dc8`: `safeFieldRules` now scopes each exception to the specific kind/parent-path where the field is actually defined (with a new regression test confirming a CRD can't smuggle a credential under `secretName`/`serviceAccountToken`), and the duplicated test case now uses `spec.secret`. Only remaining issue is the DCO check failing — commit `7775dc8` is missing its `Signed-off-by` line. Once that's amended and force-pushed, this is good to merge from my side.

**Comment by dolliecoder** [comment]
Hi @adity1raut , I am taking this over since it overlaps with my LFX mentorship project. Thank you for all the great work you've already done here

**Comment by dolliecoder** [comment]
Opened #3752 carrying these commits forward with @adity1raut as co-author, since the DCO sign-off had to come from the commit author. Same code, rebased onto current master, DCO green. Thanks again @adity1raut happy for you to close this one whenever.

**Comment by khuswant18** [comment]
@matthyx #3752 supersedes #3255 , we can close this PR
